gdal/port/cpl_userfaultfd.cpp

```diff
@@ -132,44 +149,144 @@ static void * fault_handler(void * ptr)
+    // Read page fault events
+    bytes_read = static_cast<ssize_t>(read(ctx->uffd, ctx->uffd_msgs, MAX_MESSAGES*sizeof(uffd_msg)));
+    if (bytes_read < 1) {
       if (errno == EWOULDBLOCK) continue;
       else break;
     }
 
-    fault_addr = msg.arg.pagefault.address & ~(ctx->page_size-1);
-    offset = static_cast<uint64_t>(fault_addr) - reinterpret_cast<uint64_t>(ctx->vma_ptr);
-    bytes_needed = static_cast<off_t>(ctx->file_size - offset);
-    if (bytes_needed > ctx->page_size) bytes_needed = ctx->page_size;
-
-    // Limit number of pages
-    if ((page_limit > 0) && (pages_used >= page_limit)) {
-      pages_used = 0;
-      ctx->page_ptr = mmap(ctx->page_ptr, ctx->page_size, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS|MAP_FIXED, -1, 0);
-      if (ctx->page_ptr == BAD_MMAP) {
-        ctx->page_ptr = nullptr;
-        break;
+    // If too many pages are in use, evict all pages (evict them from
+    // RAM and swap, not just to swap).  It is impossible to control
+    // which/when threads access the VMA, so access to the VMA has to
+    // forbidden while the activity is in progress.
+    //
+    // That is done by (1) installing special handlers for SIGSEGV and
+    // SIGBUS, (2) mprotecting the VMA so that any threads accessing
+    // it receive either SIGSEGV or SIGBUS (which one is apparently a
+    // function of the C library, at least on one non-Linux GNU
+    // system[1]), (3) unregistering the VMA from userfaultfd,
+    // remapping the VMA to evict the pages, registering the VMA
+    // again, (4) making the VMA accessible again, and finally (5)
+    // restoring the previous signal-handling behavior.
+    //
+    // [1] https://lists.debian.org/debian-bsd/2011/05/msg00032.html
+    pthread_mutex_lock(&mutex);
+    if ((ctx->page_limit > 0) && (ctx->pages_used > ctx->page_limit)) {
+      struct sigaction segv;
+      struct sigaction old_segv;
+      struct sigaction bus;
+      struct sigaction old_bus;
+
+      // Step 1 from the block comment above
+      segv.sa_handler = signal_handler;
+      bus.sa_handler = signal_handler;
+      if (sigaction(SIGSEGV, &segv, &old_segv) == -1) {
+      	CPLError(CE_Failure, CPLE_AppDefined,
+      		 "fault_handler: sigaction(SIGSEGV) failed");
+      	break;
+      }
+      if (sigaction(SIGBUS, &bus, &old_bus) == -1) {
+      	CPLError(CE_Failure, CPLE_AppDefined,
+      		 "fault_handler: sigaction(SIGBUS) failed");
+      	break;
       }
-    }
 
-    // Copy data into page
-    if (VSIFSeekL(file, offset, SEEK_SET)) break;
-    if (VSIFReadL(ctx->page_ptr, bytes_needed, 1, file) != 1) break;
-    pages_used++;
-
-    uffdio_copy.src = reinterpret_cast<uintptr_t>(ctx->page_ptr);
-    uffdio_copy.dst = fault_addr;
-    uffdio_copy.len = static_cast<uintptr_t>(ctx->page_size);
-    uffdio_copy.mode = 0;
-    uffdio_copy.copy = 0;
-    if (ioctl(ctx->uffd, UFFDIO_COPY, &uffdio_copy) == -1) break;
+      // WARNING LACK OF THREAD-SAFETY.
```

Comment:
I want to make sure that this question is scrutinized.  This is definitely not ideal, but I think there are are similar issues elsewhere in the system (in particular another thread changing `SIGUSR1` in between the setting and insetting of that handler in `cpl_virtualmem.cpp` might exhibit the same problem.

If the approach that you see here is used, users must be warned to never do anything that will alter `SIGSEGV` or `SIGBUS` while there are actively-faulting threads (if `GDAL_UFFD_LIMIT` is in use, if it is not in use then there is no problem).

Reply:
I should also add that this does not prevent `GDAL_UFFD_LIMIT` and `/vsimem` from being used in the same program, even from different threads.  But one must take care that there are no threads actively generating faults governed by `GDAL_UFFD_LIMIT` at the same time `/vsimem` is being setup.